### PR TITLE
Fixed: upon SIGTERM reception, send an interruption signal to children process

### PR DESCRIPTION
**Pull request by smaarn**
This partially fixes issue #746 (at least on my mac).

Test cases being performed:
- The classic `Ctrl-C` thing
- The `kill -SIGTERM <pid>` scenario

### CODE REVIEW

**Comment by smaarn** [comment]
Validated on Mac OS X 10.12 using Python 3.7.6.
On Windows, using Python 3.8.1 the Ctrl-C behaviour is still operational but there doesn't seem to be an equivalent to sending a "smooth" shutdown signal.

**Comment by morpheus65535** [comment]
And shutdown/restart button are still working in the UI?

On windows, did you take a look at `taskkill`?
https://superuser.com/questions/959364/on-windows-how-can-i-gracefully-ask-a-running-program-to-terminate

**Comment by smaarn** [comment]
It turns out that on Windows there is no need to do anything since Task kill natively kills children processes.
Running `TASKKILL /F /PID <pid> /T` already does the trick.

**Comment by smaarn** [comment]
Test cases:
- [x] Shutdown through UI - Windows
- [ ] Shutdown through UI - OS X
- [ ] Shutdown through UI - Linux
- [ ] `kill -SIGTERM <PID>` - OS X
- [ ] `kill -SIGTERM <PID>` - Linux
- [ ] `TASKKILL /F /PID <PID> /T` - Windows
- [x] Restart through UI - Windows
- [ ] Restart through UI - OS X
- [ ] Restart through UI - Linux

File: bazarr.py [hunk 1]
```diff
@@ -89,12 +128,22 @@ def daemon():
...
     
+    def shutdown(signal_no=signal.SIGINT):
+        # indicates that everything should stop
+        daemonStatus.stop()
+        daemonStatus.interrupt(signal_no)
+        # emulate a Ctrl C command on itself (bypasses the signal thing but, then, emulates the "Ctrl+C break")
+        os.kill(os.getpid(), signal_no)
     
-    start_bazarr()
+    signal.signal(signal.SIGTERM, lambda signal_no, frame: shutdown(signal.SIGINT))
```
Comment (smaarn) [inline bazarr.py]: So here the signal we send to the children processes is actually a "nice" interruption. There is an open question about whether we should actually send a more brutal one...

Comment (morpheus65535) [reply bazarr.py]: Is it possible to first send a "nice" signal then, if it fails to exit, for ce kill the process?

**Comment by morpheus65535** [comment]
Please make your PR to dev branch. We never commit directly to master.

File: bazarr.py (hunk 1 above)
Comment (smaarn) [reply bazarr.py]: Working on it.
I'll introduce a timeout of 2 minutes will that be sufficient ? (I think it should from what I've observed)

Comment (morpheus65535) [reply bazarr.py]: 2 minutes should be way enough!

**Comment by smaarn** [comment]
Test cases:

- [ ] Shutdown through UI - Windows - Python 2
- [ ] Shutdown through UI - Windows - Python 3
- [ ] Shutdown through UI - OS X - Python 2
- [ ] Shutdown through UI - OS X - Python 3
- [ ] Shutdown through UI - Linux - Python 2
- [ ] Shutdown through UI - Linux - Python 3
- [x] `kill -SIGTERM <PID>` - OS X - Python 2
- [x] `kill -SIGTERM <PID>` - OS X - Python 3
- [ ] `kill -SIGTERM <PID>` - Linux - Python 2
- [ ] `kill -SIGTERM <PID>` - Linux - Python 3
- [ ] `TASKKILL /F /PID <PID> /T` - Windows - Python 2
- [ ] `TASKKILL /F /PID <PID> /T` - Windows - Python 3
- [ ] Restart through UI - Windows - Python 2
- [ ] Restart through UI - Windows - Python 3
- [ ] Restart through UI - OS X - Python 2
- [ ] Restart through UI - OS X - Python 3
- [ ] Restart through UI - Linux - Python 2
- [ ] Restart through UI - Linux - Python 3

**Comment by smaarn** [comment]
I tried to validate it on python2... Turns out that on my NAS I cannot even install it now :|

I presume it should work similarly to Mac OS X but if someone has the will / means to test it...

**Comment by morpheus65535** [comment]
I've tested it with Python 2.7.16 on Linux but `kill -SIGTERM`dowsn't end subprocess. As we drop support for Python 2.x in 2 weeks, it could be good enough for me. If you want to fix it, that's good but I can also merge this one as it is. What do you want?

**Comment by smaarn** [comment]
@morpheus65535 I'll try having a look though since it's still supposed to be supported. At worse we could merge it once python 2 is no longer supported (would even simplify the code ;) ).

I'll try having this ironed out by sunday (I actually may have an option to rebuild python2 for my NAS come to think of it !)

**Comment by smaarn** [comment]
@morpheus65535 just tested it using Python 2.7.15 and it seemed to work.

What I did:

```
/usr/local/python/bin/python bazarr.py &
echo $!>bazarr.pid
```
Then wait for two processes to be present and then run
```
kill -SIGTERM $(cat bazarr.pid)
```
And everybody was gone... Did I miss something ?

**Comment by smaarn** [comment]
(I give you there is a limitation with my PR: there is no "Bazarr exited" printed out when sending the SIGTERM signal)

**Comment by morpheus65535** [comment]
I've just tested again with the instruction you provided and both processes still run.

**Comment by smaarn** [comment]
Though termination is not immediate this is bad. I would even qualify this as a huge regression :(
Trying to reproduce the issue here. I would personally strongly agree with not merging until one of the below conditions is met:

1. Python 2 is officially not supported (I think I can wait :D )
2. The regression cause is identified and solved

**Comment by morpheus65535** [comment]
Could you implement this new mechanism only for PY3 and leave PY2 as it is?

**Comment by smaarn** [comment]
Test cases:

- [ ] Shutdown through UI - Windows - Python 3
- [ ] `Ctrl-C` - Windows - Python 3
- [ ] `TASKKILL /F /PID <PID> /T` - Windows - Python 3
- [ ] Restart through UI - Windows - Python 3

- [ ] Shutdown through UI - Windows - Python 2
- [ ] Restart through UI - Windows - Python 2
- [ ] `TASKKILL /F /PID <PID> /T` - Windows - Python 2
- [ ] `Ctrl-C` - Windows - Python 2

- [ ] Shutdown through UI - OS X - Python 2
- [ ] Restart through UI - OS X - Python 2
- [ ] `Ctrl-C` - OS X - Python 2
- [ ] `kill -SIGTERM <PID>` - OS X - Python 2

- [ ] Shutdown through UI - OS X - Python 3
- [ ] Restart through UI - OS X - Python 3
- [ ] `Ctrl-C` - OS X - Python 3
- [ ] `kill -SIGTERM <PID>` - OS X - Python 3

- [ ] Shutdown through UI - Linux - Python 2
- [ ] Restart through UI - Linux - Python 2
- [ ] `Ctrl-C` - Linux - Python 2
- [ ] `kill -SIGTERM <PID>` - Linux - Python 2

- [ ] Shutdown through UI - Linux - Python 3
- [ ] Restart through UI - Linux - Python 3
- [ ] `Ctrl-C` - Linux - Python 3
- [ ] `kill -SIGTERM <PID>` - Linux - Python 3

**Comment by morpheus65535** [comment]
Thank you very much for this one! :-)

**Comment by smaarn** [comment]
Thanks for the patience in reviewing !

**Comment by smaarn** [comment]
> @morpheus65535 just tested it locally and to me it would seem to be related to PR #812 which changed redirections and which seem to be required here: when using pipes it seems that there would be dead locks when killing it (adding logs would actually make it explicit).
> 
> Additionally I'd say that there is also PR #810 which would be required here (it enables for proper shut down to actually occur).
> 
> I'll make a corresponding PR if that's alright with you.

Hum I think I got confused here, let me recheck.

**Pull request merged**